File: leetcode/0040.combination-sum-II/go/001.backtrack.go
func combinationSum2(candidates []int, target int) [][]int {
    res := [][]int{}

    sort.Ints(candidates)

    var backtrack func(int, []int, int)
    backtrack = func(i int, subset []int, sum int) {
        if sum == target {
            tmp := make([]int, len(subset))
            copy(tmp, subset)
            res = append(res, tmp)
            return
        }

        if i == len(candidates) || sum > target {
            return
        }

        subset = append(subset, candidates[i])
        backtrack(i+1, subset, sum + candidates[i])

        subset = subset[:len(subset)-1]

        for i+1 < len(candidates) && candidates[i] == candidates[i+1] {
            i++
        }

        backtrack(i+1, subset, sum)

    }

    backtrack(0, []int{}, 0)

    return res
}

File: leetcode/0040.combination-sum-II/rust/001.backtrack.rs
impl Solution {
    pub fn combination_sum2(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut candidates = candidates;
        candidates.sort_unstable();

        let mut res: Vec<Vec<i32>> = Vec::new();
        let mut subset: Vec<i32> = Vec:: new();

        fn backtrack(
            i: usize,
            target: i32,
            sum: i32,
            candidates: &[i32],
            subset: &mut Vec<i32>,
            res: &mut Vec<Vec<i32>>
        ) {
            if sum == target {
                res.push(subset.clone());
                return
            }
            if i >= candidates.len() || sum > target{
                return
            }

            subset.push(candidates[i]);
            backtrack(i+1, target, sum + candidates[i], candidates, subset, res);

            subset.pop();
            let mut nextI = i+1;
            while nextI < candidates.len() && candidates[i] == candidates[nextI] {
                nextI += 1;
            }
            backtrack(nextI, target, sum, candidates, subset, res);
        }

        backtrack(0, target, 0, &candidates, &mut subset, &mut res);

        res
    }
}
